File: php/src/PlutusConfig.php
<?php

declare(strict_types=1);

namespace SlaunchX\Plutus;

use Closure;
use OpenSSLAsymmetricKey;
use SlaunchX\Plutus\Exception\ConfigurationException;
use SlaunchX\Plutus\Support\Keys;
use SlaunchX\Plutus\Support\Nonce;

/**
 * SDK 配置。构造后不可变, 密钥句柄按需惰性装载并缓存。
 *
 * 四把密钥职责严格分离, 不得混用:
 * - `merchant_auth` 私钥: 对请求规范串签名 (必填);
 * - `platform_auth` 公钥: 校验响应签名与 Webhook 签名;
 * - `merchant_enc` 私钥: 解密敏感响应与 Webhook 载荷;
 * - `platform_enc` 公钥: 加密请求体。
 */
final class PlutusConfig
{
    /** @var array<string, OpenSSLAsymmetricKey> */
    private array $keyCache = [];

    /**
     * @param string        $baseUrl                     商户 API 基址, 必须是**对外 CONSUMER API 域名**
     *                                                   (如 `https://consumer-api.example.com`)。
     *                                                   不能填源站地址, 也不能自行拼接
     *                                                   `/prometheus`、`/api/v1/consumer` 等内部前缀——
     *                                                   这些改写由边缘完成, SDK 只对外部路径签名,
     *                                                   baseUrl 配错会导致签名 PATH 与实际请求路径不一致
     *                                                   而验签失败。
     * @param string        $apiKey                      `X-Api-Key` 的值 (API Key 业务 ID)
     * @param string        $merchantAuthPrivateKeyPem   商户认证私钥 (PKCS#8 PEM)
     * @param string|null   $platformAuthPublicKeyPem    平台认证公钥 (SPKI PEM), 响应验签与 Webhook 验签所需
     * @param string|null   $merchantEncPrivateKeyPem    商户加密私钥 (PKCS#8 PEM), 敏感响应与 Webhook 解密所需
     * @param string|null   $platformEncPublicKeyPem     平台加密公钥 (SPKI PEM), 请求加密所需
     * @param string|null   $platformEncKeyId            平台加密公钥指纹; 为 null 时从 PEM 推导
     * @param string|null   $platformAuthKeyId           平台认证公钥指纹; 非 null 时与响应头 `X-Platform-Signing-Key-Id` 比对
     * @param string        $apiVersion                  `X-API-VERSION` 的值, 当前恒为 `1`
     * @param bool          $verifyResponseSignature     是否校验响应签名, 默认开启
     * @param bool          $requireSignatureOnErrorResponses 非 2xx 响应缺少 `X-Response-Signature` 时是否也强制验签, 默认关闭。
     *                                                   与之无关: HTTP 2xx 缺签名头一律抛
     *                                                   {@see \SlaunchX\Plutus\Exception\ResponseSignatureException},
     *                                                   不受本开关影响; 非 2xx 缺签名头默认放行
     *                                                   (`ApiResponse::$signatureVerified === false`),
     *                                                   打开本开关后非 2xx 缺签名头同样抛异常
     * @param bool          $throwOnErrorStatus          响应判定为失败时是否抛出类型化异常, 默认开启
     * @param bool          $sendCanonicalQuery          发送规范化后的 query 而非原始串, 默认开启
     * @param int           $connectTimeoutMs            连接超时 (毫秒)
     * @param int           $timeoutMs                   整体超时 (毫秒)
     * @param string        $userAgent                   `User-Agent` 请求头
     * @param int           $webhookTimestampToleranceMs Webhook 时间戳容差 (毫秒); 0 表示不校验
     * @param Closure|null  $nonceGenerator              自定义 nonce 生成器 `fn(): string`
     * @param Closure|null  $clock                       自定义时钟 `fn(): string`, 返回 Unix 毫秒十进制串
     * @param array<int, mixed> $curlOptions             追加的 cURL 选项, 覆盖 SDK 默认值
     * @param bool          $strictEncryptedRouteValidation 加密请求的外部路径不在
     *                                                   {@see \SlaunchX\Plutus\EncryptedRoutes} 已知表中时
     *                                                   是否直接拒绝, 默认 `false` (仅记录一条 `error_log`
     *                                                   提示, 不阻断请求)。设计取舍见
     *                                                   {@see \SlaunchX\Plutus\EncryptedRoutes} 类注释。
     */
    public function __construct(
        public readonly string $baseUrl,
        public readonly string $apiKey,
        public readonly string $merchantAuthPrivateKeyPem,
        public readonly ?string $platformAuthPublicKeyPem = null,
        public readonly ?string $merchantEncPrivateKeyPem = null,
        public readonly ?string $platformEncPublicKeyPem = null,
        public readonly ?string $platformEncKeyId = null,
        public readonly ?string $platformAuthKeyId = null,
        public readonly string $apiVersion = '1',
        public readonly bool $verifyResponseSignature = true,
        public readonly bool $requireSignatureOnErrorResponses = false,
        public readonly bool $throwOnErrorStatus = true,
        public readonly bool $sendCanonicalQuery = true,
        public readonly int $connectTimeoutMs = 5000,
        public readonly int $timeoutMs = 30000,
        public readonly string $userAgent = 'slaunchx-plutus-php-sdk/1.0',
        public readonly int $webhookTimestampToleranceMs = 0,
        public readonly ?Closure $nonceGenerator = null,
        public readonly ?Closure $clock = null,
        public readonly array $curlOptions = [],
        public readonly bool $strictEncryptedRouteValidation = false,
    ) {
        if ($apiKey === '') {
            throw new ConfigurationException('apiKey 不能为空');
        }
        if ($apiVersion === '') {
            throw new ConfigurationException('apiVersion 不能为空, 当前协议恒为 "1"');
        }
        if ($merchantAuthPrivateKeyPem === '') {
            throw new ConfigurationException('merchantAuthPrivateKeyPem 不能为空');
        }
        if ($baseUrl !== '' && !preg_match('#^https?://#i', $baseUrl)) {
            throw new ConfigurationException('baseUrl 必须以 http:// 或 https:// 开头');
        }
    }

    /**
     * 商户认证私钥句柄。
     */
    public function merchantAuthPrivateKey(): OpenSSLAsymmetricKey
    {
        return $this->keyCache['merchant_auth'] ??= Keys::loadPrivateKey(
            $this->merchantAuthPrivateKeyPem,
            'merchant_auth 私钥'
        );
    }

    /**
     * 平台认证公钥句柄。
     *
     * @throws ConfigurationException 未配置 `platformAuthPublicKeyPem`
     */
    public function platformAuthPublicKey(): OpenSSLAsymmetricKey
    {
        if ($this->platformAuthPublicKeyPem === null) {
            throw new ConfigurationException('未配置 platformAuthPublicKeyPem, 无法校验响应或 Webhook 签名');
        }

        return $this->keyCache['platform_auth'] ??= Keys::loadPublicKey(
            $this->platformAuthPublicKeyPem,
            'platform_auth 公钥'
        );
    }

    /**
     * 商户加密私钥句柄。
     *
     * @throws ConfigurationException 未配置 `merchantEncPrivateKeyPem`
     */
    public function merchantEncPrivateKey(): OpenSSLAsymmetricKey
    {
        if ($this->merchantEncPrivateKeyPem === null) {
            throw new ConfigurationException('未配置 merchantEncPrivateKeyPem, 无法解密敏感响应或 Webhook');
        }

        return $this->keyCache['merchant_enc'] ??= Keys::loadPrivateKey(
            $this->merchantEncPrivateKeyPem,
            'merchant_enc 私钥'
        );
    }

    /**
     * 平台加密公钥句柄。
     *
     * @throws ConfigurationException 未配置 `platformEncPublicKeyPem`
     */
    public function platformEncPublicKey(): OpenSSLAsymmetricKey
    {
        if ($this->platformEncPublicKeyPem === null) {
            throw new ConfigurationException('未配置 platformEncPublicKeyPem, 无法加密请求体');
        }

        return $this->keyCache['platform_enc'] ??= Keys::loadPublicKey(
            $this->platformEncPublicKeyPem,
            'platform_enc 公钥'
        );
    }

    /**
     * `X-Platform-Encryption-Key-Id` 的值: 平台加密公钥指纹。
     *
     * 优先使用显式配置的 `platformEncKeyId`, 否则从平台加密公钥 PEM 推导。
     */
    public function resolvePlatformEncKeyId(): string
    {
        if ($this->platformEncKeyId !== null && $this->platformEncKeyId !== '') {
            return $this->platformEncKeyId;
        }

        return Keys::fingerprint($this->platformEncPublicKey());
    }

    /**
     * 商户加密公钥指纹, 用于与信封 `keyFingerprint` 比对。
     */
    public function merchantEncKeyFingerprint(): string
    {
        return Keys::fingerprint($this->merchantEncPrivateKey());
    }

    /**
     * 商户认证公钥指纹, 用于与平台登记值核对。
     */
    public function merchantAuthKeyFingerprint(): string
    {
        return Keys::fingerprint($this->merchantAuthPrivateKey());
    }

    /**
     * 当前时间戳, Unix 毫秒十进制字符串。
     */
    public function currentTimestampMillis(): string
    {
        if ($this->clock !== null) {
            return (string) ($this->clock)();
        }

        return (string) (int) round(microtime(true) * 1000);
    }

    /**
     * 生成一次性 nonce。
     */
    public function generateNonce(): string
    {
        if ($this->nonceGenerator !== null) {
            return (string) ($this->nonceGenerator)();
        }

        return Nonce::generate();
    }
}

File: java/src/main/java/com/slaunchx/plutus/sdk/PlutusConfig.java
package com.slaunchx.plutus.sdk;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.slaunchx.plutus.sdk.crypto.PemKeys;
import com.slaunchx.plutus.sdk.exception.PlutusConfigurationException;

import java.net.URI;
import java.net.http.HttpClient;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.time.Duration;

/**
 * SDK 配置。
 *
 * <p>四把密钥职责严格分离,不得混用:
 * <ul>
 *   <li>{@code merchant_auth} <b>私钥</b>:对请求规范串签名。必填。</li>
 *   <li>{@code platform_auth} <b>公钥</b>:验证响应签名与 Webhook 签名。
 *       {@link Builder#verifyResponseSignature(boolean)} 为 {@code true}(默认)时必填。</li>
 *   <li>{@code merchant_enc} <b>私钥</b>:解密敏感响应与 Webhook 载荷。使用相应功能时必填。</li>
 *   <li>{@code platform_enc} <b>公钥</b>:加密请求体。调用加密端点时必填。</li>
 * </ul>
 */
public final class PlutusConfig {

    private final ProtocolProfile protocolProfile;
    private final URI baseUrl;
    private final String apiKey;
    private final String apiVersion;
    private final RSAPrivateKey merchantAuthPrivateKey;
    private final RSAPublicKey platformAuthPublicKey;
    private final RSAPrivateKey merchantEncPrivateKey;
    private final RSAPublicKey merchantEncPublicKey;
    private final RSAPublicKey platformEncPublicKey;
    private final String merchantEncFingerprint;
    private final String platformEncFingerprint;
    private final boolean verifyResponseSignature;
    private final boolean requireSignatureOnErrorResponses;
    private final boolean strictEncryptedRouteValidation;
    private final NonceGenerator nonceGenerator;
    private final HttpClient httpClient;
    private final Duration requestTimeout;
    private final ObjectMapper objectMapper;

    private PlutusConfig(Builder b) {
        if (b.protocolProfile == null) throw new PlutusConfigurationException("protocolProfile is required");
        this.protocolProfile = b.protocolProfile;
        if (b.apiKey == null || b.apiKey.isBlank()) {
            throw new PlutusConfigurationException("缺少 apiKey");
        }
        if (b.merchantAuthPrivateKeyPem == null) {
            throw new PlutusConfigurationException("缺少 merchantAuthPrivateKeyPem");
        }
        if (b.apiVersion == null || b.apiVersion.isBlank()) {
            throw new PlutusConfigurationException("apiVersion 不可为空");
        }
        this.baseUrl = b.baseUrl == null ? null : URI.create(stripTrailingSlash(b.baseUrl));
        this.apiKey = b.apiKey;
        this.apiVersion = b.apiVersion;
        this.merchantAuthPrivateKey = PemKeys.parsePrivateKey(b.merchantAuthPrivateKeyPem);
        this.platformAuthPublicKey = b.platformAuthPublicKeyPem == null
                ? null : PemKeys.parsePublicKey(b.platformAuthPublicKeyPem);
        this.merchantEncPrivateKey = b.merchantEncPrivateKeyPem == null
                ? null : PemKeys.parsePrivateKey(b.merchantEncPrivateKeyPem);
        this.platformEncPublicKey = b.platformEncPublicKeyPem == null
                ? null : PemKeys.parsePublicKey(b.platformEncPublicKeyPem);
        this.verifyResponseSignature = b.verifyResponseSignature;
        this.requireSignatureOnErrorResponses = b.requireSignatureOnErrorResponses;
        this.strictEncryptedRouteValidation = b.strictEncryptedRouteValidation;
        if (this.verifyResponseSignature && this.platformAuthPublicKey == null) {
            throw new PlutusConfigurationException(
                    "启用响应验签时必须提供 platformAuthPublicKeyPem; 如确需关闭请显式调用 verifyResponseSignature(false)");
        }

        RSAPublicKey merchantEncPublic = null;
        if (b.merchantEncPublicKeyPem != null) {
            merchantEncPublic = PemKeys.parsePublicKey(b.merchantEncPublicKeyPem);
        } else if (this.merchantEncPrivateKey != null) {
            merchantEncPublic = PemKeys.derivePublicKey(this.merchantEncPrivateKey);
        }
        this.merchantEncPublicKey = merchantEncPublic;
        this.merchantEncFingerprint = merchantEncPublic == null ? null : PemKeys.fingerprint(merchantEncPublic);
        this.platformEncFingerprint = this.platformEncPublicKey == null
                ? null : PemKeys.fingerprint(this.platformEncPublicKey);

        this.nonceGenerator = b.nonceGenerator == null ? NonceGenerator.secureRandomHex() : b.nonceGenerator;
        this.requestTimeout = b.requestTimeout == null ? Duration.ofSeconds(30) : b.requestTimeout;
        this.objectMapper = b.objectMapper == null ? new ObjectMapper() : b.objectMapper;
        this.httpClient = b.httpClient == null
                ? HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build()
                : b.httpClient;
    }

    private static String stripTrailingSlash(String url) {
        String trimmed = url.trim();
        while (trimmed.endsWith("/")) {
            trimmed = trimmed.substring(0, trimmed.length() - 1);
        }
        return trimmed;
    }

    /**
     * @return 新的构造器
     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * 必须是<b>对外 CONSUMER API 域名</b>,不能是源站地址,也不能自行拼接
     * {@code /prometheus}、{@code /api/v1/consumer} 等边缘/源站前缀——边缘(Cloudflare/nginx)
     * 负责把外部路径改写为源站内部路径,SDK 与商户都不应感知该改写。正确示例:
     * {@code https://<consumer-api-host>}(配合 {@code path} 传 {@code /card-products/xxx});
     * 错误示例:{@code https://origin-host/prometheus/api/v1/consumer}(签名必然失败)。
     *
     * @return 商户 API 基地址,不含路径;仅在使用 {@link PlutusClient} 时必填
     */
    public ProtocolProfile protocolProfile() { return protocolProfile; }

    public URI baseUrl() {
        return baseUrl;
    }

    /** @return API Key 业务 ID */
    public String apiKey() {
        return apiKey;
    }

    /** @return 契约主版本,默认 {@code 1} */
    public String apiVersion() {
        return apiVersion;
    }

    /** @return 商户认证私钥 */
    public RSAPrivateKey merchantAuthPrivateKey() {
        return merchantAuthPrivateKey;
    }

    /** @return 平台认证公钥;未配置时为 {@code null} */
    public RSAPublicKey platformAuthPublicKey() {
        return platformAuthPublicKey;
    }

    /** @return 商户加密私钥;未配置时为 {@code null} */
    public RSAPrivateKey merchantEncPrivateKey() {
        return merchantEncPrivateKey;
    }

    /** @return 商户加密公钥(显式配置或从私钥推导);未配置时为 {@code null} */
    public RSAPublicKey merchantEncPublicKey() {
        return merchantEncPublicKey;
    }

    /** @return 平台加密公钥;未配置时为 {@code null} */
    public RSAPublicKey platformEncPublicKey() {
        return platformEncPublicKey;
    }

    /** @return 商户加密公钥指纹,用于敏感响应 AAD 与信封 {@code keyFingerprint} 校验 */
    public String merchantEncFingerprint() {
        return merchantEncFingerprint;
    }

    /** @return 平台加密公钥指纹,用于 {@code X-Platform-Encryption-Key-Id} 与加密请求 AAD */
    public String platformEncFingerprint() {
        return platformEncFingerprint;
    }

    /** @return 是否强制验证响应签名,默认 {@code true} */
    public boolean verifyResponseSignature() {
        return verifyResponseSignature;
    }

    /**
     * 非 2xx 响应缺签名头时是否也强制验签,默认 {@code false}。
     *
     * <p>在 {@link #verifyResponseSignature()} 为 {@code true} 的前提下,响应缺少
     * {@code X-Response-Signature} 时的默认策略是:HTTP 2xx 一律抛
     * {@link com.slaunchx.plutus.sdk.exception.PlutusSignatureException};非 2xx 放行,
     * {@link PlutusResponse#signatureVerified()} 为 {@code false}。本开关打开后,非 2xx
     * 缺签名头同样抛验签异常。
     *
     * @return 严格要求错误响应携带签名返回 {@code true}
     */
    public boolean requireSignatureOnErrorResponses() {
        return requireSignatureOnErrorResponses;
    }

    /**
     * 加密端点的 routeTemplate(即请求外部路径)是否必须命中
     * {@link com.slaunchx.plutus.sdk.crypto.EncryptedRoutes} 已知表,默认 {@code false}。
     *
     * <p>关闭(默认)时,未登记路由只触发一次 WARNING 级日志,不阻断请求——避免平台新增
     * 加密端点后, 尚未升级 SDK 常量表的商户请求集体失败。打开后, 未登记路由直接抛
     * {@link com.slaunchx.plutus.sdk.exception.PlutusException},已登记路由不受影响。
     *
     * @return 是否启用加密端点路由严格校验
     */
    public boolean strictEncryptedRouteValidation() {
        return strictEncryptedRouteValidation;
    }

    /** @return nonce 生成器 */
    public NonceGenerator nonceGenerator() {
        return nonceGenerator;
    }

    /** @return HTTP 客户端 */
    public HttpClient httpClient() {
        return httpClient;
    }

    /** @return 单次请求超时 */
    public Duration requestTimeout() {
        return requestTimeout;
    }

    /** @return JSON 序列化器 */
    public ObjectMapper objectMapper() {
        return objectMapper;
    }

    /**
     * {@link PlutusConfig} 构造器。密钥以 PEM 文本传入,由 SDK 解析并校验。
     */
    public static final class Builder {

        private ProtocolProfile protocolProfile = ProtocolProfile.REQUEST_BOUND_V1;

        public Builder protocolProfile(ProtocolProfile value) { this.protocolProfile = value; return this; }

        private String baseUrl;
        private String apiKey;
        private String apiVersion = "1";
        private String merchantAuthPrivateKeyPem;
        private String platformAuthPublicKeyPem;
        private String merchantEncPrivateKeyPem;
        private String merchantEncPublicKeyPem;
        private String platformEncPublicKeyPem;
        private boolean verifyResponseSignature = true;
        private boolean requireSignatureOnErrorResponses = false;
        private boolean strictEncryptedRouteValidation = false;
        private NonceGenerator nonceGenerator;
        private HttpClient httpClient;
        private Duration requestTimeout;
        private ObjectMapper objectMapper;

        private Builder() {
        }

        /**
         * 必须是<b>对外 CONSUMER API 域名</b>,不能是源站地址,也不能自行拼接
         * {@code /prometheus}、{@code /api/v1/consumer} 等边缘/源站前缀。正确示例:
         * {@code baseUrl("https://<consumer-api-host>")},配合
         * {@code PlutusRequest.post("/card-products/10010106/shared/cards/create")}；
         * 错误示例:{@code baseUrl("https://origin-host/prometheus/api/v1/consumer")}
         * ——边缘会再次改写路径, 签名必然失败。
         *
         * @param baseUrl 商户 API 基地址,如 {@code https://consumer-api.example.com};末尾斜杠会被去除
         * @return 自身
         */
        public Builder baseUrl(String baseUrl) {
            this.baseUrl = baseUrl;
            return this;
        }

        /**
         * @param apiKey API Key 业务 ID,写入 {@code X-Api-Key}
         * @return 自身
         */
        public Builder apiKey(String apiKey) {
            this.apiKey = apiKey;
            return this;
        }

        /**
         * @param apiVersion 契约主版本,默认 {@code 1};参与签名,不可为空
         * @return 自身
         */
        public Builder apiVersion(String apiVersion) {
            this.apiVersion = apiVersion;
            return this;
        }

        /**
         * @param pem {@code merchant_auth} PKCS#8 私钥 PEM
         * @return 自身
         */
        public Builder merchantAuthPrivateKeyPem(String pem) {
            this.merchantAuthPrivateKeyPem = pem;
            return this;
        }

        /**
         * @param pem {@code platform_auth} SPKI 公钥 PEM
         * @return 自身
         */
        public Builder platformAuthPublicKeyPem(String pem) {
            this.platformAuthPublicKeyPem = pem;
            return this;
        }

        /**
         * @param pem {@code merchant_enc} PKCS#8 私钥 PEM
         * @return 自身
         */
        public Builder merchantEncPrivateKeyPem(String pem) {
            this.merchantEncPrivateKeyPem = pem;
            return this;
        }

        /**
         * 可选。不提供时从 {@code merchant_enc} 私钥推导,用于计算自身指纹。
         *
         * @param pem {@code merchant_enc} SPKI 公钥 PEM
         * @return 自身
         */
        public Builder merchantEncPublicKeyPem(String pem) {
            this.merchantEncPublicKeyPem = pem;
            return this;
        }

        /**
         * @param pem {@code platform_enc} SPKI 公钥 PEM
         * @return 自身
         */
        public Builder platformEncPublicKeyPem(String pem) {
            this.platformEncPublicKeyPem = pem;
            return this;
        }

        /**
         * @param verify 是否强制验证响应签名,默认 {@code true};关闭意味着放弃响应完整性保护
         * @return 自身
         */
        public Builder verifyResponseSignature(boolean verify) {
            this.verifyResponseSignature = verify;
            return this;
        }

        /**
         * 非 2xx 响应缺签名头时是否也强制验签,默认 {@code false}。
         *
         * <p>响应缺少 {@code X-Response-Signature} 时的默认策略:HTTP 2xx 一律抛
         * {@link com.slaunchx.plutus.sdk.exception.PlutusSignatureException},不把未验证的数据
         * 交给业务代码;非 2xx 默认放行,{@link PlutusResponse#signatureVerified()} 为
         * {@code false},调用方仍可读到平台返回的错误码。打开本开关后,非 2xx 缺签名头也抛
         * 验签异常。
         *
         * <p>本开关只在 {@link #verifyResponseSignature(boolean)} 为 {@code true} 时生效。
         *
         * @param require 是否严格要求错误响应携带签名,默认 {@code false}
         * @return 自身
         */
        public Builder requireSignatureOnErrorResponses(boolean require) {
            this.requireSignatureOnErrorResponses = require;
            return this;
        }

        /**
         * 加密端点的 routeTemplate 是否必须命中
         * {@link com.slaunchx.plutus.sdk.crypto.EncryptedRoutes} 已知表,默认 {@code false}。
         *
         * <p>默认(非严格)模式下,未登记路由只触发一次 WARNING 级日志提示,不阻断请求——
         * 平台新增加密端点是可预期的演进,若直接拒绝会让尚未升级 SDK 常量表的商户请求集体
         * 失败。打开严格模式后,未登记路由直接抛
         * {@link com.slaunchx.plutus.sdk.exception.PlutusException},已登记路由不受影响;
         * 适合需要提前发现路径拼写错误或平台端点变更的商户。
         *
         * @param strict 是否启用严格校验,默认 {@code false}
         * @return 自身
         */
        public Builder strictEncryptedRouteValidation(boolean strict) {
            this.strictEncryptedRouteValidation = strict;
            return this;
        }

        /**
         * @param generator 自定义 nonce 生成器,默认为 32 字符十六进制
         * @return 自身
         */
        public Builder nonceGenerator(NonceGenerator generator) {
            this.nonceGenerator = generator;
            return this;
        }

        /**
         * @param httpClient 自定义 JDK HTTP 客户端
         * @return 自身
         */
        public Builder httpClient(HttpClient httpClient) {
            this.httpClient = httpClient;
            return this;
        }

        /**
         * @param timeout 单次请求超时,默认 30 秒
         * @return 自身
         */
        public Builder requestTimeout(Duration timeout) {
            this.requestTimeout = timeout;
            return this;
        }

        /**
         * @param mapper 自定义 Jackson 实例
         * @return 自身
         */
        public Builder objectMapper(ObjectMapper mapper) {
            this.objectMapper = mapper;
            return this;
        }

        /**
         * @return 配置实例
         * @throws PlutusConfigurationException 必填项缺失或密钥不合规时抛出
         */
        public PlutusConfig build() {
            return new PlutusConfig(this);
        }
    }
}
